### DISCUSSION TITLE
HTML API: Upstream bugs/issues

### DISCUSSION THREAD

**Discussion by dmsnell** [WordPress/gutenberg#57803] [HTML API]
## Browsers

<h3 id="cdata-handling-inside-foreign-elements"><code>CDATA</code> handling inside foreign elements.</h3>

The HTML5 specification indicates that CDATA sections exist within foreign elements and provides an example.

 - `CDATA` can [only exist within foreign elements](https://html.spec.whatwg.org/#cdata-sections). 
 - There is [no provision](https://html.spec.whatwg.org/#parsing-main-inforeign) for `CDATA` inside foreign elements.
 - https://github.com/whatwg/html/issues/4016

```html
<p>You can add a string to a number, but this stringifies the number:</p>
<math>
 <ms><![CDATA[x<y]]></ms>
 <mo>+</mo>
 <mn>3</mn>
 <mo>=</mo>
 <ms><![CDATA[x<y3]]></ms>
</math>
```

In Firefox this renders as expected, but in Safari and in Chrome the `CDATA` sections are hidden as comments, which is how they would be parsed in normal HTML content.

To highlight this discrepancy we can invert the comparison inside the `CDATA` section to `<![CDATA[x>y3]]>`. Now, in Firefox this still renders properly, but in Safari and Chrome we have `y3]]>` as text inside the Math render, because they are ending the `CDATA` section at the first `>` as if it were a bogus comment.

| Firefox | Safari |
|---|---|
| <img width="500" alt="Screenshot 2024-01-12 at 8 13 17 AM" src="https://github.com/WordPress/gutenberg/assets/5431237/283fd491-d6ef-4b76-a4cf-f5438eb43cee"> |  <img width="567" alt="Screenshot 2024-01-12 at 8 13 47 AM" src="https://github.com/WordPress/gutenberg/assets/5431237/8ebf1c57-951e-48ed-bb65-9126124d2f90"> |

**Comment by sirreal** [comment]
I understand that CDATA sections should be respected only in foreign content, which are only SVG and MathML (`<svg>` and `<math>`).

I've tested Chrome (120), Safary (17) and Firefox (121). You can use these data-uris below for handy testing.

CDATA outside of foreign content should be treated as bogus comments, ending at the first `>`. All browsers render `y]]>` here:

```
data:text/html,<body><![CDATA[x > y]]>
```

This (svg) seems to render correctly in all browsers, `x > y`:

```
data:text/html,<svg><text y="20"><![CDATA[x > y]]>
```

This (mathml) seems to render correctly in Firefox (`x > y`) but not in Chrome or Safari (`y]]>`).

```
data:text/html,<math><mtext><![CDATA[x > y]]>
```

It seems like [this blink issue](https://bugs.chromium.org/p/chromium/issues/detail?id=881983&q=cdata%20mathml&can=2) may be related.

I think Firefox is doing the correct thing according to the specification?

**Reply by dmsnell** [reply]
> If there is an [adjusted current node](https://html.spec.whatwg.org/multipage/parsing.html#adjusted-current-node) and it is not an element in the [HTML namespace](https://infra.spec.whatwg.org/#html-namespace), then switch to the [CDATA section state](https://html.spec.whatwg.org/multipage/parsing.html#cdata-section-state). Otherwise, this is a [cdata-in-html-content](https://html.spec.whatwg.org/multipage/parsing.html#parse-error-cdata-in-html-content) [parse error](https://html.spec.whatwg.org/multipage/parsing.html#parse-errors)

https://html.spec.whatwg.org/multipage/parsing.html#markup-declaration-open-state

This was hiding in plain sight, and we probably saw this when implementing it before, but since we don't support the foreign elements yet we skipped over it.
